akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala

```diff
@@ -754,7 +754,7 @@ import akka.util.OptionVal
   }
 
   private def shortCircuitBatch(): Unit = {
-    while (!shortCircuitBuffer.isEmpty && currentLimit > 0 && activeInterpreters.nonEmpty) shortCircuitBuffer
+    while (!shortCircuitBuffer.isEmpty && currentLimit > 0 && (activeInterpreters.nonEmpty || newShells.nonEmpty)) shortCircuitBuffer
```

Comment:
Did you check if removing the `activeInterpreters.nonEmpty` clause has a similar effect?

Reply:
not sure I'm following, 
but as far as I understand it: in the common case newShells will be empty and activeInterpreters won't. 
so in most case the short circuit buffer holds events relevant for the active interpreters and it doesn't make any sense to drop it.

Reply:
I meant that using `while (!shortCircuitBuffer.isEmpty && currentLimit > 0)` might suffice. AFAICS the `&& activeInterpreters.nonEmpty` is an optimization that avoids dispatching internal signals that have been scheduled before all interpreters were shutdown. You could say that this optimization was the reason for the bug here because it also prevented running the loop in the forgotten case you found here.

Reply:
@johanandren linked me to #19299, TBH I haven't followed the entire discussion but it seems the team had good reasons to come up with these optimizations.
when approaching such a delicate and thoughtful code, my approach is to apply the minimal required modification in order to avoid ill effects.

re. the condition and optimization, I generally suspect that in most invocations of this method activeInterpreters will not be empty and newShells will, so the condition will short-circuit. so I believe both my modifications and your suggested modification will have negligible if any effect on performance.

looking at the second condition in this method, it seems to have been designed for cases not covered by the loop, notice it also fail for the scenario in the issue:
```
if (!shortCircuitBuffer.isEmpty && currentLimit == 0) self ! Resume
```

so it might make sense to simplify both conditions in order to guarantee full coverage.

Reply:
> so it might make sense to simplify both conditions in order to guarantee full coverage.

Yes, I agree, we could at least check if all the tests pass with simplified conditions.

Reply:
👍

Reply:
@jrudolph , I've just pushed a simplified version of the condition.
however, I think this can be further simplified by converting the `while` into a `@tailrec` method. this way it'd have to exit points, one when the queue is drained, the other when `currentLimit` is reached (with a side effect of sending `Resume` to `self`). I think it'd also be simpler as the existing two current conditions will actually be handled in the same chain (of conditions).
WDUT?

Reply:
FlatMapMergeBenchmark doesn't seem to show any regression with this change. As you say it's all a bit delicate, but AFAICS it looks good. I'm not sure a `@tailrec` variation would be much simpler, but might be worth a try?